### ISSUE TITLE
New Rule: no-destructing

### ISSUE THREAD

**Issue by Apacheli** [eslint/eslint#12569]
**Please describe what the rule should do:**
This rule would disallow users from using the destructing syntax.

**What category of rule is this? (place an "X" next to just one item)**
[ ] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[ ] Enforces code style (layout)
[X] Other (please specify: ES6 syntax, and a bit of 2 and 3)

**Provide 2-3 code examples that this rule will warn about:**
```js
// Looks weird and confusing
const [,,,,,,,,,,,,,,,,,,,,yes] = [1, 2, 3];

// Unrealistic case, but funny behavior
delete Array.prototype[Symbol.iterator];
const [x, y] = [1, 2];
```
```js
// Polluting the namespace
const { Animal, Bat, Cat, Dog, Elephant, Fish, Giraffe } = require('./animals');

// Cleaner, better intellisense, and less invasive to the namespace
const animals = require('./animals');
animals.Animal;
```
```js
// Verbose renaming
const { veryLongNameIsVeryLong: short } = require('./long-names');

// Better alternative than the above
const short = require('./long-names').veryLongNameIsVeryLong;
```

**Why should this rule be included in ESLint (instead of a plugin)?**
I believe that this rule would allow for better code consistency for those who are using ES6 syntax, which ESlint supports. I also think that this rule could possibly be merged with `prefer-destructing` as they are both similar rules.

**Are you willing to submit a pull request to implement this rule?**
I wish I could, but I might not get there on time. I'd appreciate if someone else could implement it. If nobody does, then that's all good. I can still implement it.

**Comment by kaicataldo** [comment]
I don’t think this is a good match for core, but it actually is already enforceable using [`no-restricted-syntax`](https://eslint.org/docs/rules/no-restricted-syntax).

**Comment by Apacheli** [comment]
Hello, @kaicataldo, could you please show me an example of disallow destructing with that rule? I was almost able to achieve expected behavior, however the following is considered an error:
```js
/* eslint no-restricted-syntax: ["error", "VariableDeclarator"] */

let { x } = { x: 5 }
// Errors at y
let y = 10
```

**Comment by kaicataldo** [comment]
Sure thing!

I believe this should allow you to enforce this:

```js
/* eslint no-restricted-syntax: ["error",
  { "selector": "ObjectPattern", "message": "Object destructuring is not allowed." },
  { "selector": "ArrayPattern", "message": "Array destructuring is not allowed." },
] */

const { foo } = bar;
const [a, b, c] = baz;
```

Example using [our demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXJlc3RyaWN0ZWQtc3ludGF4OiBbXCJlcnJvclwiLFxuICB7IFwic2VsZWN0b3JcIjogXCJPYmplY3RQYXR0ZXJuXCIsIFwibWVzc2FnZVwiOiBcIk9iamVjdCBkZXN0cnVjdHVyaW5nIGlzIG5vdCBhbGxvd2VkLlwiIH0sXG4gIHsgXCJzZWxlY3RvclwiOiBcIkFycmF5UGF0dGVyblwiLCBcIm1lc3NhZ2VcIjogXCJBcnJheSBkZXN0cnVjdHVyaW5nIGlzIG5vdCBhbGxvd2VkLlwiIH0sXG5dICovXG5cbmNvbnN0IHsgZm9vIH0gPSBiYXI7XG5jb25zdCBbYSwgYiwgY10gPSBiYXo7Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjo2LCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7fSwiZW52Ijp7fX19).

Hope that helps!

**Comment by Apacheli** [comment]
Thank you for your help. I was not aware of this rule before I opened this issue. I appreciate it!
